### PR TITLE
Feat/configurable fullscreen element

### PR DESCRIPTION
**Pull request by cjpillsbury**
This PR adds the ability to target something other than the `<media-controller/>` instance for fullscreen. It also adds an example that takes advantage of this.

TODO:
- [x] Update Docs

### CODE REVIEW

File: examples/standalone-controls.html [hunk 1]
```diff
@@ -26,7 +26,7 @@
       <h1>Standalone Controls Example</h1>
       <p>
         This example shows how individual controls can be used without the
-        <code>media-container</code> element.
+        <code>media-controller</code> element.
```
Comment (cjpillsbury) [inline examples/standalone-controls.html:29]: This change is unrelated. Just noticed that this still referenced `media-container`.

File: src/js/media-controller.js [hunk 2]
```diff
@@ -551,6 +552,17 @@ class MediaController extends MediaContainer {
...
+  get fullscreenElement() {
+    if (this.hasAttribute('fullscreen-element')) {
+      return document.getElementById(this.getAttribute('fullscreen-element'));
+    }
+    return this.#fullscreenElement ?? this;
+  }
+
+  set fullscreenElement(element) {
+    this.#fullscreenElement = element;
+  }
```
Comment (gkatsev) [inline src/js/media-controller.js:564]: I would assume that if this property is set, it should unset the attribute/getter should check the private var before checking the attribute.

**Review by dylanjha — CHANGES_REQUESTED** [review]
Missing `docs/` update for the new `fullscreen-element` attribute?

File: src/js/media-controller.js [hunk 3]
```diff
@@ -551,6 +552,17 @@ class MediaController extends MediaContainer {
     this.enableHotkeys();
   }
 
+  get fullscreenElement() {
+    if (this.hasAttribute('fullscreen-element')) {
+      return document.getElementById(this.getAttribute('fullscreen-element'));
```
Comment (gkatsev) [inline src/js/media-controller.js:557]: I think this is probably good enough for the majority usecase. If someone has nested WCs, they can use the setter.

**Comment by cjpillsbury** [comment]
> Missing `docs/` update for the new `fullscreen-element` attribute?

lol added a todo to the description already (good timing).

File: src/js/media-controller.js (hunk 2 above)
Comment (cjpillsbury) [reply src/js/media-controller.js:564]: Yeah I held off on that for a couple of reasons:

1. I was trying to avoid @luwes's common "attribute sprouting" concerns here.
2. It wasn't obvious to me that if someone sets and then unsets the value that they'd want the `fullscreen-element` attribute to "disappear into the aether"

Originally I just had an `attributeChangedCallback` case that would explicitly set the `fullscreenElement`, but I ended up removing that to allow for the separation and precendence of the different values that I implemented here.

That said, I don't feel strongly about this impl, and it's easy enough to change. If we want to treat them as "synonymous", I think we would need to re-add my `attributeChangedCallback` logic and do what you're proposing.

File: examples/wrapper.html [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        top: 0;
+        left: 0;
+        right: 0;
+        bottom: 0;
+        pointer-events: none;
+        color: #ffffff;
+      }
+
+      #wrapper:is(:fullscreen, :-webkit-full-screen, :-moz-full-screen, :-ms-fullscreen) {
+        --fullscreen-overaly-display: 'initial';
```
Comment (luwes) [inline examples/wrapper.html:43]:
```suggestion
        --fullscreen-overlay-display: 'initial';
```

File: examples/wrapper.html [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        flex-shrink: 1;
+      }
+
+      #wrapper > .spacer {
+        background-color: var(--media-background-color, #000);
+        flex-grow: 1;
+      }
+
+      #wrapper > .fullscreen-overlay {
+        display: var(--fullscreen-overaly-display, none);
```
Comment (luwes) [inline examples/wrapper.html:31]:
```suggestion
        display: var(--fullscreen-overlay-display, none);
```

File: src/js/media-controller.js [hunk 6]
```diff
@@ -48,6 +48,7 @@ class MediaController extends MediaContainer {
   }
 
   #hotKeys = new AttributeTokenList(this, 'hotkeys');
+  #fullscreenElement;
```
Comment (luwes) [inline src/js/media-controller.js:51]: I have a little concern about the naming here. will fullscreen be the only feature that needs a settable element?

Comment (luwes) [reply src/js/media-controller.js:51]: yes, came back to me. the resize observer will also have to be set to this element right?

File: examples/wrapper.html (hunk 4 above)
Comment (cjpillsbury) [reply examples/wrapper.html:43]: whoops that should also not be a string (I make this mistake all the time with CSS, esp. CSS vars. @luwes Wanna update your suggestion to include fixing not one, but two of my silly mistakes?

Comment (luwes) [reply examples/wrapper.html:43]: updated

File: src/js/media-controller.js (hunk 6 above)
Comment (cjpillsbury) [reply src/js/media-controller.js:51]: @luwes Yeah I was struggling with a name here, too (names are hard). I definitely don't feel confident that there isn't a future world where there isn't either:
1. a single prop/attr that treats many things as the "media container substitute" (we've discussed some of these in the past, e.g. gestures and keypresses)
2. multiple props that are for setting elements for different purposes on a `media-controller`

Because of that, I didn't want to be overly presumptuous about the name right now (since, at least for case 1, that would be a breaking change). That said, I also don't like the specific name, so if you have an alternative proposal that doesn't assume too much of what it _**could**_ be used for in the future, I'm down.

File: src/js/media-controller.js (hunk 3 above)
Comment (cjpillsbury) [reply src/js/media-controller.js:557]: Yeah, at the very least, it feels like something that could be a followup.

File: src/js/media-controller.js (hunk 6 above)
Comment (dylanjha) [reply src/js/media-controller.js:51]: `#fullscreenElement` seemed okay to me since this is a narrow use case, but if we're looking for something that would also apply to the resize observer `#containerElement`?

Then again, there could be use cases where you want the fullscreenElement to be one thing, and some other element to be another thing and not be forced to make them the same thing. For that reason, I'd go with `#fullscreenElement`

Comment (luwes) [reply src/js/media-controller.js:51]: true, it's hard to say at the moment. if the rumors are true we'll have at least pipElement soon too.

it's not it will be a breaking change I think. 

we could add containerElement later as well and make it clear that it's the element of the full application in any presentation mode.

File: src/js/media-controller.js [hunk 7]
```diff
@@ -592,6 +604,13 @@ class MediaController extends MediaContainer {
         this.#hotKeys.value = newValue;
     } else if (attrName === 'default-stream-type') {
       this.propagateMediaState(MediaUIAttributes.MEDIA_STREAM_TYPE);
+    } else if (attrName === 'fullscreen-element') {
+      const el = newValue 
+        ? document.getElementById('fullscreen-element')
```
Comment (luwes) [inline src/js/media-controller.js:609]:
```suggestion
        ? this.getRootNode()?.getElementById(newValue) 
```

- could be in a shadow DOM so best to get it from the shadow root where media controller is in 
- fixes the element id string

File: examples/wrapper.html [hunk 8]
```diff
@@ -0,0 +1,116 @@
+<!DOCTYPE html>
```
Comment (heff) [inline examples/wrapper.html:1]: I think we're trying to push more things to docs rather than examples, right? Would this example fit under a guide? Can it be demoed in docs easily enough?

File: examples/wrapper.html [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+<html lang="en">
+  <head>
+    <meta name="viewport" content="width=device-width" />
+    <title>Media Chrome Standalone Controls Example</title>
```
Comment (heff) [inline examples/wrapper.html:5]: New title needed

File: src/js/media-controller.js [hunk 10]
```diff
@@ -592,6 +604,14 @@ class MediaController extends MediaContainer {
         this.#hotKeys.value = newValue;
     } else if (attrName === 'default-stream-type') {
       this.propagateMediaState(MediaUIAttributes.MEDIA_STREAM_TYPE);
+    } else if (attrName === 'fullscreen-element') {
```
Comment (heff) [inline src/js/media-controller.js:607]: We've really gotta nail down this dash/no-dash thing 😆 (not a blocker)

File: src/js/media-controller.js [hunk 11]
```diff
@@ -592,6 +604,14 @@ class MediaController extends MediaContainer {
...
+      const el = newValue 
+        // @ts-ignore
+        ? this.getRootNode()?.getElementById(newValue) 
+        : undefined;
```
Comment (heff) [inline src/js/media-controller.js:611]: Non-blocker: I think `null` tends to be more common for "intentionally empty". We end up with a potential mix of undefined and null negative values as is (e.g. `getElementById(notAnElement) == null`).

File: src/js/media-controller.js [hunk 12]
```diff
@@ -592,6 +604,14 @@ class MediaController extends MediaContainer {
...
+      // NOTE: Setting the internal private prop here to not
+      // clear the attribute that was just set (CJP).
+      this.#fullscreenElement = el;
```
Comment (heff) [inline src/js/media-controller.js:614]: Probably overkill, but I'm wondering if we should exit fullscreen on any existing real element when changing this value.

**Review by heff — APPROVED** [review]
Love it. 👍

**Review by dylanjha — APPROVED** [review]

File: src/js/media-controller.js (hunk 12 above)
Comment (cjpillsbury) [reply src/js/media-controller.js:614]: I think it should be fine and I kinda like that this doesn't have any direct visible side effects. I know I'd be annoyed if I had to worry about my programmatic workflow unavoidably having a user experience impact.
